experimental/ast/token.go

```diff
@@ -0,0 +1,641 @@
+		return Span{}
+	}
+
+	if !t.IsLeaf() {
+		start, end := t.StartEnd()
+		a, _ := start.offsets()
+		_, b := end.offsets()
+
+		return t.Context().NewSpan(a, b)
+	}
+
+	return t.Context().NewSpan(t.offsets())
+}
+
+// StartEnd returns the open and close tokens for this token.
+//
+// If this is a leaf token, start and end will be the same token and will compare as equal.
+//
+// Panics if this is a nil token.
+func (t Token) StartEnd() (start, end Token) {
+	t.panicIfNil()
+
+	switch impl := t.impl(); {
+	case impl == nil:
+		switch synth := t.synthetic(); {
+		case synth.IsLeaf():
+			return t, t
+		case synth.IsOpen():
+			start = t
+			end = synth.otherEnd.With(t)
+		case synth.IsClose():
+			start = synth.otherEnd.With(t)
+			end = t
+		}
+
+	case impl.IsLeaf():
+		return t, t
+	case impl.IsOpen():
+		start = t
+		end = (t.raw + rawToken(impl.Offset())).With(t)
+	case impl.IsClose():
+		start = (t.raw + rawToken(impl.Offset())).With(t)
+		end = t
+	}
+
+	return
+}
+
+// Offsets returns the byte offsets of this token within the file it came from.
+//
+// The return value for synthetic tokens is unspecified.
+//
+// Note that this DOES NOT include any child tokens!
+func (t Token) offsets() (start, end int) {
+	if t.IsSynthetic() {
+		return
+	}
+
+	end = int(t.impl().end)
+	// If this is the first token, the start is implicitly zero.
+	if t.raw == 1 {
+		return 0, end
+	}
+
+	prev := (t.raw - 1).With(t)
+	return int(prev.impl().end), end
+}
+
+// Children returns a Cursor over the children of this token.
+//
+// If the token is nil or is a leaf token, returns nil.
+func (t Token) Children() *Cursor {
+	if t.Nil() || t.IsLeaf() {
+		return nil
+	}
+
+	if impl := t.impl(); impl != nil {
+		start, end := t.StartEnd()
+		return &Cursor{
+			withContext: t.withContext,
+			start:       start.raw + 1, // Skip the start!
+			end:         end.raw,
+		}
+	}
+
+	synth := t.synthetic()
+	if synth.IsClose() {
+		return synth.otherEnd.With(t).Children()
+	}
+	return &Cursor{
+		withContext: t.withContext,
+		stream:      synth.children,
+	}
+}
+
+// Name converts this token into its corresponding identifier name, potentially
+// performing normalization.
+//
+// Currently, we perform no normalization, so this is the same value as Text(), but
+// that may change in the future.
+//
+// Returns "" for non-identifiers.
+func (t Token) Name() string {
+	if t.Kind() != TokenIdent {
+		return ""
+	}
+	return t.Text()
+}
+
+// AsUInt converts this token into an unsigned integer if it is a numeric token.
+// bits is the maximum number of bits that are used to represent this value.
+//
+// Otherwise, or if the result would overflow, returns 0, false.
+func (t Token) AsInt() (uint64, bool) {
+	if t.Kind() != TokenNumber {
+		return 0, false
+	}
+
+	// Check if this number has already been parsed for us.
+	vAny, present := t.Context().literals[t.raw]
+	if v, ok := vAny.(uint64); present && ok {
+		return v, true
+	}
+
+	// Otherwise, it's an base 10 integer.
```

Comment:
So hex and octal formats are parsed by... the lexer?

Reply:
Yup. This is explained in detail in the comment on `Context.literals`.